### PR TITLE
feat: support `bigquery.vector_search()`

### PR DESCRIPTION
**Pull request by ashleyxuu**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes internal #344019437 🦕

### CODE REVIEW

File: bigframes/bigquery/__init__.py [hunk 1]
```diff
@@ -20,19 +20,17 @@
...
 
-import typing
+from typing import Literal, Optional, Union
 
+import bigframes.bigquery.utils as utils
 import bigframes.constants as constants
 import bigframes.core.groupby as groupby
 import bigframes.operations as ops
 import bigframes.operations.aggregations as agg_ops
+import bigframes.pandas as bpd
```
Comment (tswast) [inline bigframes/bigquery/__init__.py:30]: Cloud RAD is really bad at finding classes when we do things this way. Let's use `bigframes.series.Series` and `bigframes.dataframe.DataFrame` for our type annotations, please. See: b/296390934

File: bigframes/bigquery/utils.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# Copyright 2024 Google LLC
```
Comment (tswast) [inline bigframes/bigquery/utils.py:1]: No `utils` files, please (go/no-utils). We should consolidate our SQL generation following Trevor's recent refactor. See: https://github.com/googleapis/python-bigquery-dataframes/pull/682

**Review by tswast — CHANGES_REQUESTED** [review]

File: bigframes/bigquery/__init__.py (hunk 1 above)
Comment (ashleyxuu) [reply bigframes/bigquery/__init__.py:30]: Done.

File: bigframes/bigquery/utils.py (hunk 2 above)
Comment (ashleyxuu) [reply bigframes/bigquery/utils.py]: Done.

File: bigframes/core/sql.py [hunk 3]
```diff
@@ -169,3 +169,47 @@ def ordering_clause(
...
+
+    if query_column_to_search is not None:
+        query_str = f"""
+    SELECT
+        query.*,
+        base.*,
+        distance,
+    FROM VECTOR_SEARCH(
+        TABLE `{base_table}`,
+        "{column_to_search}",
```
Comment (tswast) [inline bigframes/core/sql.py:194]: I think it's possible for column names to have quote characters.

Let's use `simple_literal()` to make sure we do the right escaping for all of the options values https://github.com/googleapis/python-bigquery-dataframes/blob/ea5d367d5ecc6826d30082e75c957af8362c9e61/bigframes/core/sql.py#L36

Comment (tswast) [reply bigframes/core/sql.py:194]: Even if column names don't allow such characters, we don't want `vector_search()` to accidentally expose our users to SQL injection attacks.

File: bigframes/bigquery/__init__.py [hunk 4]
```diff
@@ -148,3 +152,150 @@ def array_to_string(series: series.Series, delimiter: str) -> series.Series:
...
+    Conduct vector search which searches embeddings to find semantically similar entities.
+
+    **Examples:**
+
+
+        >>> import bigframes.pandas as bpd
+        >>> import bigframes.bigquery as bbq
+        >>> bpd.options.display.progress_bar = None
+
+    DataFrame embeddings for which to find nearest neighbors:
```
Comment (tswast) [inline bigframes/bigquery/__init__.py:178]: Let's clarify here that the `ARRAY<FLOAT>` column is used as the search query.

File: bigframes/bigquery/__init__.py [hunk 5]
```diff
@@ -148,3 +152,150 @@ def array_to_string(series: series.Series, delimiter: str) -> series.Series:
...
+        ...             top_k=2)
+             embedding  id my_embedding  distance
+        dog    [1. 2.]   1      [1. 2.]       0.0
+        cat  [3.  5.2]   5    [5.  5.4]  2.009975
+        dog    [1. 2.]   4    [1.  3.2]       1.2
+        cat  [3.  5.2]   2      [2. 4.]   1.56205
+        <BLANKLINE>
+        [4 rows x 4 columns]
+
+    You can specify the name of the column in the query DataFrame embeddings and distance type:
```
Comment (tswast) [inline bigframes/bigquery/__init__.py:213]: Presumably `query_column_to_search` is required in this case, right? Let's tell the user that if so. Or if not, what happens if they don't specify?

**Review by tswast — CHANGES_REQUESTED** [review]

File: bigframes/bigquery/__init__.py (hunk 5 above)
Comment (ashleyxuu) [reply bigframes/bigquery/__init__.py]: done

File: bigframes/bigquery/__init__.py (hunk 4 above)
Comment (ashleyxuu) [reply bigframes/bigquery/__init__.py]: Done

File: bigframes/bigquery/__init__.py [hunk 6]
```diff
@@ -148,3 +152,153 @@ def array_to_string(series: series.Series, delimiter: str) -> series.Series:
...
+    Conduct vector search which searches embeddings to find semantically similar entities.
+
+    **Examples:**
+
+
+        >>> import bigframes.pandas as bpd
+        >>> import bigframes.bigquery as bbq
+        >>> bpd.options.display.progress_bar = None
+
+    DataFrame embeddings for which to find nearest neighbors, and ARRAY<FLOAT> column
```
Comment (tswast) [inline bigframes/bigquery/__init__.py:178]: Let's use code font for `ARRAY<FLOAT>`. Also the BQ type is called `FLOAT64` in standard SQL https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#floating_point_types

```suggestion
    DataFrame embeddings for which to find nearest neighbors. The ``ARRAY<FLOAT64>`` column
```

**Review by tswast — APPROVED** [review]

**Pull request merged**